// secret-cipher/dist/index.js
import { n as e, r as t } from "./chunk-Dqa2HsxW.js";
//#region src/runtimes/types.ts
var n = {
	NODE: "node",
	WEB: "web",
	WEB_STANDARDS: "web-standards"
}, r = {
	ENVIRONMENT_UNAVAILABLE: "CR:environment-unavailable",
	INVALID_BASE64URL: "CR:invalid-base64url",
	INVALID_CIPHERTEXT: "CR:invalid-ciphertext",
	INVALID_KEY: "CR:invalid-key",
	INVALID_NONCE: "CR:invalid-nonce",
	INVALID_RUNTIME: "CR:invalid-runtime",
	INVALID_URL: "CR:invalid-url",
	OPERATION_FAILED: "CR:operation-failed"
}, i = {
	AUTO: "auto",
	...n
}, a = { SEPARATOR: "." }, o = { DEFAULT_BASE_PATH: "/s" }, s = {
	AES_KEY: 32,
	IV: 12,
	SALT: 16
}, c = { ACCESS: 32 }, l = { PREFIX: "secret-cipher:" }, u = class extends Error {
	code;
	constructor(e, t, n) {
		super(t, n), this.name = "CipherError", this.code = e;
	}
}, d = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_", f = /^[A-Za-z0-9_-]*$/, p = (e) => {
	let t = "", n = 0;
	for (; n + 2 < e.length; n += 3) {
		let r = e[n] << 16 | e[n + 1] << 8 | e[n + 2];
		t += d[r >> 18 & 63], t += d[r >> 12 & 63], t += d[r >> 6 & 63], t += d[r & 63];
	}
	let r = e.length - n;
	if (r === 1) {
		let r = e[n] << 16;
		t += d[r >> 18 & 63], t += d[r >> 12 & 63];
	}
	if (r === 2) {
		let r = e[n] << 16 | e[n + 1] << 8;
		t += d[r >> 18 & 63], t += d[r >> 12 & 63], t += d[r >> 6 & 63];
	}
	return t;
}, m = (e) => {
	if (!f.test(e) || e.length % 4 == 1) throw new u(r.INVALID_BASE64URL, "Invalid base64url value.");
	let t = 0, n = 0, i = [];
	for (let a of e) {
		let e = d.indexOf(a);
		if (e < 0) throw new u(r.INVALID_BASE64URL, "Invalid base64url value.");
		t = t << 6 | e, n += 6, n >= 8 && (n -= 8, i.push(t >> n & 255));
	}
	return Uint8Array.from(i);
}, h = (e) => {
	let t = globalThis.TextEncoder;
	if (typeof t != "function") throw new u(r.ENVIRONMENT_UNAVAILABLE, "TextEncoder is unavailable in this environment.");
	return new t().encode(e);
}, g = (e) => {
	let t = globalThis.TextDecoder;
	if (typeof t != "function") throw new u(r.ENVIRONMENT_UNAVAILABLE, "TextDecoder is unavailable in this environment.");
	return new t().decode(e);
}, _ = (e) => typeof e == "object" && !!e, v = (e, t) => typeof e[t] == "function", ee = (e) => {
	if (!_(e) || !v(e, "getRandomValues")) return !1;
	let t = e.subtle;
	return _(t) ? v(t, "digest") && v(t, "importKey") && v(t, "deriveBits") && v(t, "encrypt") && v(t, "decrypt") : !1;
}, y = (e) => {
	if (!Number.isSafeInteger(e) || e <= 0) throw TypeError("byteLength must be a positive safe integer.");
}, b = (e = Reflect.get(globalThis, "crypto"), t = n.WEB_STANDARDS) => {
	if (!ee(e)) return null;
	let { subtle: r } = e;
	return {
		name: t,
		randomBytes(t) {
			return y(t), e.getRandomValues(new Uint8Array(t));
		},
		async sha256(e) {
			return new Uint8Array(await r.digest("SHA-256", e));
		},
		async hkdfSha256({ key: e, salt: t, info: n, byteLength: i }) {
			y(i);
			let a = await r.importKey("raw", e, "HKDF", !1, ["deriveBits"]), o = await r.deriveBits({
				name: "HKDF",
				hash: "SHA-256",
				salt: t,
				info: n
			}, a, i * 8);
			return new Uint8Array(o);
		},
		async aesGcmEncrypt({ key: e, iv: t, plaintext: n, additionalData: i }) {
			let a = await r.importKey("raw", e, "AES-GCM", !1, ["encrypt"]), o = await r.encrypt({
				name: "AES-GCM",
				iv: t,
				additionalData: i,
				tagLength: 128
			}, a, n);
			return new Uint8Array(o);
		},
		async aesGcmDecrypt({ key: e, iv: t, ciphertext: n, additionalData: i }) {
			let a = await r.importKey("raw", e, "AES-GCM", !1, ["decrypt"]), o = await r.decrypt({
				name: "AES-GCM",
				iv: t,
				additionalData: i,
				tagLength: 128
			}, a, n);
			return new Uint8Array(o);
		}
	};
}, x = () => {
	let e = Reflect.get(globalThis, "window");
	return typeof e != "object" || !e ? null : b(Reflect.get(e, "crypto"), n.WEB);
}, te = () => {
	let e = Reflect.get(globalThis, "process");
	if (typeof e != "object" || !e) return !1;
	let t = Reflect.get(e, "versions");
	return typeof t != "object" || !t ? !1 : typeof Reflect.get(t, "node") == "string";
}, ne = () => Reflect.has(globalThis, "window"), S = async () => {
	let { createNodeRuntime: e } = await Promise.resolve().then(() => Se);
	return e();
}, C = async (e) => e === i.WEB ? x() : e === i.WEB_STANDARDS ? b() : e === i.NODE ? S() : null, re = async () => {
	if (ne()) {
		let e = x();
		if (e) return e;
	}
	if (te()) {
		let e = await S();
		if (e) return e;
	}
	return b() || null;
}, w = async ({ runtime: e, preferredRuntime: t = i.AUTO } = {}) => {
	if (e) return e;
	let n = t === i.AUTO ? await re() : await C(t);
	if (n) return n;
	throw new u(r.ENVIRONMENT_UNAVAILABLE, `No ${t} cipher runtime is available in this environment.`);
}, T = (e = o.DEFAULT_BASE_PATH) => (e.startsWith("/") ? e : `/${e}`).replace(/\/+$/, ""), ie = (e) => e ? e.replace(/\/+$/, "") : "", E = ({ readId: e, origin: t, basePath: n, fragment: r }) => {
	let i = `${T(n)}/${encodeURIComponent(e)}`;
	return `${ie(t)}${i}${r}`;
}, D = ({ input: e, basePath: t, decodeFragment: n }) => {
	let r = oe(e), i = n(r.hash);
	if (!i) return null;
	let a = T(t);
	if (!r.pathname.startsWith(`${a}/`)) return null;
	let o = decodeURIComponent(r.pathname.slice(a.length + 1));
	return o ? {
		...i,
		readId: o
	} : null;
}, ae = (e) => /^[a-z][a-z\d+\-.]*:\/\//i.test(e), oe = (e) => {
	if (!ae(e)) throw new u(r.INVALID_URL, "Access URL must include an origin.");
	try {
		return new URL(e);
	} catch (e) {
		throw new u(r.INVALID_URL, "Invalid access URL.", { cause: e });
	}
}, se = (e) => {
	let [t, n, i, o] = e.split(a.SEPARATOR);
	if (!t || !n || !i || typeof o == "string") throw new u(r.INVALID_CIPHERTEXT, "Ciphertext must use the salt.iv.ciphertext payload format.");
	return {
		salt: m(t),
		iv: m(n),
		ciphertext: m(i)
	};
}, ce = ({ salt: e, iv: t, ciphertext: n }) => `${p(e)}.${p(t)}.${p(n)}`, O = async ({ info: e, salt: t, secret: n, runtime: r }) => r.hkdfSha256({
	key: n,
	salt: t,
	info: e,
	byteLength: s.AES_KEY
}), k = (e) => {
	let t = m(e);
	if (t.byteLength === c.ACCESS) return t;
	throw new u(r.INVALID_KEY, `Expected ${c.ACCESS} bytes.`);
}, A = (e) => (k(e), e), j = async (e = {}) => p((await w(e)).randomBytes(c.ACCESS)), M = h(`${l.PREFIX}text-key`), N = h(`${l.PREFIX}text`), P = async (e = {}) => j(e), F = (e) => A(e), le = async (e, t = {}) => {
	let n = await w(t), r = t.secret ? F(t.secret) : await P({ runtime: n }), i = k(r), a = n.randomBytes(s.SALT), o = n.randomBytes(s.IV), c = await O({
		info: M,
		salt: a,
		secret: i,
		runtime: n
	});
	return {
		cipher: ce({
			salt: a,
			iv: o,
			ciphertext: await n.aesGcmEncrypt({
				key: c,
				iv: o,
				plaintext: h(e),
				additionalData: N
			})
		}),
		secret: r
	};
}, ue = async ({ cipher: e, secret: t, runtime: n, preferredRuntime: i }) => {
	let a = await w({
		runtime: n,
		preferredRuntime: i
	}), o = se(e), s = await O({
		info: M,
		salt: o.salt,
		secret: k(t),
		runtime: a
	});
	try {
		return g(await a.aesGcmDecrypt({
			key: s,
			iv: o.iv,
			ciphertext: o.ciphertext,
			additionalData: N
		}));
	} catch (e) {
		throw new u(r.OPERATION_FAILED, "Unable to decrypt text.", { cause: e });
	}
}, I = (e) => `#${F(e)}`, L = (e) => {
	let t = e.startsWith("#") ? e.slice(1) : e;
	if (!t) return null;
	try {
		return { secret: F(t) };
	} catch {
		return null;
	}
}, de = ({ readId: e, origin: t, basePath: n, secret: r }) => E({
	readId: e,
	origin: t,
	basePath: n,
	fragment: I(r)
}), fe = (e, { basePath: t } = {}) => D({
	input: e,
	basePath: t,
	decodeFragment: L
}), R = (e, t) => {
	if (!(Number.isSafeInteger(e) && e > 0)) throw new u(r.OPERATION_FAILED, `${t} must be a positive integer.`);
}, z = (e, t) => {
	if (!(Number.isSafeInteger(e) && e >= 0)) throw new u(r.OPERATION_FAILED, `${t} must be a non-negative integer.`);
}, B = 8 * 1024 * 1024, V = h(`${l.PREFIX}file-key`), H = h(`${l.PREFIX}file-manifest-key`), U = h(`${l.PREFIX}file-chunk-nonce`), W = h(`${l.PREFIX}file-manifest`), pe = B, me = (e) => {
	if (he(e) && typeof e.name == "string" && typeof e.type == "string" && K(e.size) && G(e.chunkSize) && G(e.chunkCount) && (e.lastModified === void 0 || K(e.lastModified))) return {
		name: e.name,
		type: e.type,
		size: e.size,
		lastModified: e.lastModified,
		chunkSize: e.chunkSize,
		chunkCount: e.chunkCount
	};
	throw new u(r.INVALID_CIPHERTEXT, "Invalid file manifest.");
}, he = (e) => typeof e == "object" && !!e, G = (e) => typeof e == "number" && Number.isSafeInteger(e) && e > 0, K = (e) => typeof e == "number" && Number.isSafeInteger(e) && e >= 0, ge = async (e = {}) => j(e), q = (e) => A(e), J = (e) => `#${q(e)}`, Y = (e) => {
	let t = e.startsWith("#") ? e.slice(1) : e;
	if (!t) return null;
	try {
		return { secret: q(t) };
	} catch {
		return null;
	}
}, _e = ({ readId: e, origin: t, basePath: n, secret: r }) => E({
	readId: e,
	origin: t,
	basePath: n,
	fragment: J(r)
}), ve = (e, { basePath: t } = {}) => D({
	input: e,
	basePath: t,
	decodeFragment: Y
}), ye = async ({ manifest: e, secret: t, salt: n, runtime: r, preferredRuntime: i }) => {
	let a = await w({
		runtime: r,
		preferredRuntime: i
	}), o = n ? m(n) : a.randomBytes(s.SALT), c = a.randomBytes(s.IV), l = await O({
		info: H,
		salt: o,
		secret: k(t),
		runtime: a
	}), u = await a.aesGcmEncrypt({
		key: l,
		iv: c,
		plaintext: h(JSON.stringify(e)),
		additionalData: W
	});
	return {
		salt: p(o),
		iv: p(c),
		cipher: p(u)
	};
}, X = async ({ salt: e, iv: t, cipher: n, secret: i, runtime: a, preferredRuntime: o }) => {
	let s = await w({
		runtime: a,
		preferredRuntime: o
	}), c = await O({
		info: H,
		salt: m(e),
		secret: k(i),
		runtime: s
	});
	try {
		let e = await s.aesGcmDecrypt({
			key: c,
			iv: m(t),
			ciphertext: m(n),
			additionalData: W
		});
		return me(JSON.parse(g(e)));
	} catch (e) {
		throw new u(r.OPERATION_FAILED, "Unable to decrypt file manifest.", { cause: e });
	}
}, be = async ({ chunk: e, chunkIndex: t, chunkCount: n, chunkSize: r = B, salt: i, secret: a, runtime: o, preferredRuntime: c }) => {
	z(t, "chunkIndex"), R(n, "chunkCount"), R(r, "chunkSize");
	let l = await w({
		runtime: o,
		preferredRuntime: c
	}), u = m(i), d = k(a), f = await O({
		info: V,
		salt: u,
		secret: d,
		runtime: l
	}), p = await l.hkdfSha256({
		key: d,
		salt: u,
		info: U,
		byteLength: s.IV
	});
	return l.aesGcmEncrypt({
		key: f,
		iv: Q(p, t),
		plaintext: e,
		additionalData: Z({
			chunkIndex: t,
			chunkCount: n,
			chunkSize: r,
			plaintextLength: e.byteLength
		})
	});
}, xe = async ({ ciphertext: e, chunkIndex: t, chunkCount: n, plaintextLength: r, chunkSize: i = B, salt: a, secret: o, runtime: c, preferredRuntime: l }) => {
	z(t, "chunkIndex"), R(n, "chunkCount"), R(i, "chunkSize"), z(r, "plaintextLength");
	let u = await w({
		runtime: c,
		preferredRuntime: l
	}), d = m(a), f = k(o), p = await O({
		info: V,
		salt: d,
		secret: f,
		runtime: u
	}), h = await u.hkdfSha256({
		key: f,
		salt: d,
		info: U,
		byteLength: s.IV
	});
	return u.aesGcmDecrypt({
		key: p,
		iv: Q(h, t),
		ciphertext: e,
		additionalData: Z({
			chunkIndex: t,
			chunkCount: n,
			chunkSize: i,
			plaintextLength: r
		})
	});
}, Z = ({ chunkIndex: e, chunkCount: t, chunkSize: n, plaintextLength: r }) => h(`${l.PREFIX}file-chunk:${e}:${t}:${n}:${r}`), Q = (e, t) => {
	if (e.byteLength !== s.IV) throw new u(r.INVALID_NONCE, `Expected ${s.IV} bytes.`);
	let n = new Uint8Array(e), i = BigInt(t);
	for (let e = s.IV - 1; e >= 0; --e) n[e] ^= Number(i & 255n), i >>= 8n;
	return n;
}, Se = /* @__PURE__ */ e({ createNodeRuntime: () => $ }), $ = async () => {
	let e = b(Reflect.get(globalThis, "crypto"), n.NODE);
	if (e) return e;
	try {
		return b((await import("./__vite-browser-external-BZzOzNFq.js").then((e) => /* @__PURE__ */ t(e.default, 1))).webcrypto, n.NODE);
	} catch {
		return null;
	}
};
//#endregion
export { u as CipherError, r as ERRORS, i as RUNTIME_PREFERENCE, m as base64UrlToBytes, p as bytesToBase64Url, ge as createFileSecret, $ as createNodeRuntime, P as createTextSecret, x as createWebRuntime, b as createWebStandardsRuntime, Y as decodeFileAccessFragment, ve as decodeFileAccessUrl, L as decodeTextAccessFragment, fe as decodeTextAccessUrl, pe as defaultFileChunkSizeBytes, J as encodeFileAccessFragment, _e as encodeFileAccessUrl, I as encodeTextAccessFragment, de as encodeTextAccessUrl, q as fileSecretFromString, xe as openFileChunk, X as openFileManifest, ue as openText, w as resolveCipherRuntime, be as sealFileChunk, ye as sealFileManifest, le as sealText, F as textSecretFromString };
